src/module.c

```diff
@@ -10846,10 +10998,407 @@ sds genModulesInfoString(sds info) {
+ * errno is set:
+ * * EINVAL: The provided flags are invalid for the registration.
+ * * EALREADY: The provided configuration name is already used. */
+int RM_RegisterStringConfig(RedisModuleCtx *ctx, const char *name, RedisModuleString *default_val, unsigned int flags, RedisModuleConfigGetStringFunc getfn, RedisModuleConfigSetStringFunc setfn, RedisModuleConfigApplyFunc applyfn, void *privdata) {
+    RedisModule *module = ctx->module;
+    if (moduleConfigValidityCheck(module, name, flags, REDISMODULE_CONFIG_STRING)) {
+        return REDISMODULE_ERR;
+    }
+    ModuleConfig newStringConfig = createModuleConfig(sdsnew(name), REDISMODULE_CONFIG_STRING, applyfn, privdata, module);
+    addDefaultAndCallbacksToConfig(newStringConfig, getfn, setfn, string);
+    addModuleConfigtoList(module, &newStringConfig);
+    flags = maskModuleConfigFlags(flags);
+    addModuleStringConfig(module->name, name, flags, module->module_configs->tail->value, default_val ? default_val->ptr : NULL);
+    return REDISMODULE_OK;
+}
+
+/* 
+ Create an enum config that server clients can interact with via the 
+ * `CONFIG SET`, `CONFIG GET`, and `CONFIG REWRITE` commands. 
+ * enum_values is a array of c-strings that are valid set options. We match
+ * enum names with their index partner in int_vals.
+ * Example Implementation:
+        const char *enum_vals[3] = {"first", "second", "third"};
+        const int int_vals[3] = {0, 2, 4};
+        int enum_val = 0;
+
+        int getEnumConfigCommand(const char *name, void *privdata) {
+            return enum_val;
+        }
+        
+        int setEnumConfigCommand(const char *name, int val, void *privdata, const char **err) {
+            enum_val = val;
+            return REDISMODULE_OK;
+        }
+        ...
+        RedisModule_RegisterEnumConfig(ctx, "enum", 0, REDISMODULE_CONFIG_DEFAULT, enum_vals, int_vals, 3, getEnumConfigCommand, setEnumConfigCommand, NULL, NULL);
+
+ * See RedisModule_RegisterStringConfig for detailed general information about configs. */
+int RM_RegisterEnumConfig(RedisModuleCtx *ctx, const char *name, int default_val, unsigned int flags, const char **enum_values, const int *int_values, int num_enum_vals, RedisModuleConfigGetEnumFunc getfn, RedisModuleConfigSetEnumFunc setfn, RedisModuleConfigApplyFunc applyfn, void *privdata) {
+    RedisModule *module = ctx->module;
+    if (moduleConfigValidityCheck(module, name, flags, REDISMODULE_CONFIG_ENUM)) {
+        return REDISMODULE_ERR;
+    }
+    ModuleConfig newEnumConfig = createModuleConfig(sdsnew(name), REDISMODULE_CONFIG_ENUM, applyfn, privdata, module);
+    addDefaultAndCallbacksToConfig(newEnumConfig, getfn, setfn, enum);
+    configEnum *enum_vals = zmalloc(num_enum_vals * sizeof(configEnum) + 1);
+    for (int i = 0; i < num_enum_vals; i++) {
+        configEnum enum_config = {.name = zstrdup(enum_values[i]), .val = int_values[i]};
+        memcpy(enum_vals + i, &enum_config, sizeof(configEnum));    
+    }
```

Comment:
Probably more readable in this context.
```suggestion
    for (int i = 0; i < num_enum_vals; i++) {
        enum_vals[i]->name = zstrdup(enum_vaues[i]);
        enum_vals[i]->val = int_values[i];
    }
```